File: selected_coding_interview/codes/python/lc_207_course_schedule_s1.py
"""
File: lc_207_course_schedule_s1.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
from collections import deque
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegrees = [0 for _ in range(numCourses)]
        adjacency = [[] for _ in range(numCourses)]
        queue = deque()
        # Get the indegree and adjacency of every course.
        for cur, pre in prerequisites:
            indegrees[cur] += 1
            adjacency[pre].append(cur)
        # Get all the courses with the indegree of 0.
        for i in range(len(indegrees)):
            if not indegrees[i]: queue.append(i)
        # BFS TopSort.
        while queue:
            pre = queue.popleft()
            numCourses -= 1
            for cur in adjacency[pre]:
                indegrees[cur] -= 1
                if not indegrees[cur]: queue.append(cur)
        return not numCourses


# ======= Test Case =======
test_input_numCourses = 2
test_input_prerequisites = [[1, 0]]
expected_output = True

# ====== Driver Code ======
slt = Solution()
result = slt.canFinish(test_input_numCourses, test_input_prerequisites)
print(result)

File: selected_coding_interview/codes/java/lc_207_course_schedule/lc_207_course_schedule_s1.java
/*
* File: lc_207_course_schedule_s1.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_207_course_schedule;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public boolean canFinish(int numCourses, int[][] prerequisites) {
            int[] indegrees = new int[numCourses];
            List<List<Integer>> adjacency = new ArrayList<>();
            Queue<Integer> queue = new LinkedList<>();
            for(int i = 0; i < numCourses; i++)
                adjacency.add(new ArrayList<>());
            // Get the indegree and adjacency of every course.
            for(int[] cp : prerequisites) {
                indegrees[cp[0]]++;
                adjacency.get(cp[1]).add(cp[0]);
            }
            // Get all the courses with the indegree of 0.
            for(int i = 0; i < numCourses; i++)
                if(indegrees[i] == 0) queue.add(i);
            // BFS TopSort.
            while(!queue.isEmpty()) {
                int pre = queue.poll();
                numCourses--;
                for(int cur : adjacency.get(pre))
                    if(--indegrees[cur] == 0) queue.add(cur);
            }
            return numCourses == 0;
        }
    }

public class lc_207_course_schedule_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        int test_input_numCourses = 2;
        int[][] test_input_prerequisites = new int[][]{{1, 0}};
        var expected_output = true;

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.canFinish(test_input_numCourses, test_input_prerequisites);
        System.out.println(result);

    }
}
